feat(helm,tiller): improve list output, support filters

technosophos opened the pull request:
This commit brings `helm list` up to spec:
- Sort is done on the server
- Filter is implemented as a server operation
- Limit and max are processed after filter and sort
- Documentation is corrected to reflect the changes
- Tests added for sorting and filtering

technosophos committed:
feat(tiller): sort list in tiller, not helm

technosophos committed:
feat(helm,tiller): add support for list filters

This adds support for filtering list results. Filter strings are
passed from Helm to Tiller, where they are compiled as regular
expressions and executed against the list of releases. Only matching
releases are returned.

Filters are applied before limits and sorts.

vaikas-google commented on _proto/hapi/services/tiller.proto:
```diff
@@ -69,16 +69,33 @@ service ReleaseService {
+	// Offset is the zero-based offset at which the returned release list begins.
 	int64 offset = 2;
+
+  // SortBy is the sort field that the ListReleases server should sort data before returning.
```
Seems like the indent is off here?

vaikas-google commented on _proto/hapi/services/tiller.proto:
```diff
@@ -69,16 +69,33 @@ service ReleaseService {
+  enum SortBy {
+    UNKNOWN = 0;
+    NAME = 1;
+    LAST_RELEASED = 2;
```
Thoughts on offering the ability to do ascend/descending sort?

vaikas-google commented on _proto/hapi/services/tiller.proto:
```diff
@@ -69,16 +69,33 @@ service ReleaseService {
-	
-	// The zero-based offset at which the returned release list begins 
+
+	// Offset is the zero-based offset at which the returned release list begins.
```
Rather than have an index here, it might be better to provide a way to specify the starting 'key', if it was only say names, you could offer pagination that way, rather than have to keep track off the offset. Just a thought.

technosophos commented:
I agree with @vaikas-google about the pagination method. I'll change that. I think I will also add sort direction.

technosophos committed:
feat(helm,tiller): add list reverse, string offset

technosophos commented:
New features:

- offset is now a string (last item seen)
- `helm list -r` reverses the sort order

vaikas-google commented on _proto/hapi/services/tiller.proto:
```diff
@@ -53,27 +53,64 @@ service ReleaseService {
 
-	// The zero-based offset at which the list is positioned
-	int64 offset = 2;
+	// Offset is the last-seen release name, used to offset the next set.
```
You could combine offset / more, by renaming offset to nextKey or something like that. If it's filled in on the response, there's more available, if not, then it's empty. Not a big deal, but a thought.

technosophos replied:
So in that version, is it the name of the first value not returned? (Right now, it's the name of the last value returned.)

vaikas-google replied:
Yeah, precisely. So you'd basically fetch the limit/max +1, and if there's more results, you'd return the next item if there are more, or none if there's no more to be fetched. I think it's cleaner in that instead of having both a bool / nextToken, you could roll them into one return value.

technosophos replied:
Okay. I think I'll do that.

technosophos replied:
So I implemented it this way, but I'm not sure the result is the best user experience. So if it doesn't work well, we can switch back.

Right now, the CLI prints out a list of releases, and then, counterintuitively, a `next: RELEASE_NAME` that is not on the list just displayed, and the output doesn't match the formatting for the rest.

```
$ bin/helm list -m 3
joking-kudu
kindred-lion
left-ladybug
	next: quoting-octopus
```

vaikas-google replied:
That's fair (if it doesn't work, we'll switch back). If it's just a matter of formatting the output, that's probably easy to address, but from the pure API point, I think it's cleaner this way :) Wouldn't be the first time (nor the last) that I've been wrong though...

technosophos committed:
fix(helm,tiller): change list offset to next result

Previously, paging used the last release name of the current set to ask
for more results. Now switched to using the first name of the next set.

Not sure I like this method. It makes the user experience more
complicated.

vaikas-google commented on cmd/tiller/release_server_test.go:
```diff
@@ -217,6 +217,86 @@ func TestListReleases(t *testing.T) {
+		t.Errorf("Unexpected sort order: %v.", mrs.val.Releases)
+	}
+	if mrs.val.Releases[1].Name != "neuron" {
+		t.Errorf("Unexpected sort order: %v.", mrs.val.Releases)
```
Would it be better to use %#v in these cases, I think it contains more verbose information about the types being dumped.
